Revising VTF Modding. (#67)
R2Northstar/ModdingDocs: docs/source/guides/VTFModding.rst

VTF Overview
------------

VTF, short for "Valve Texture Format", is a texture type used by Valve in the source engine and is occasionally in Titanfall. vtf's are used for fx, animation, and other textures. 

VMT Overview
------------

VMT, short for "Valve Material Type", is a text `material <https://developer.valvesoftware.com/wiki/Material>`__ system that dictates how the game perceives a vtf outside of how it looks. It uses `parameters <https://developer.valvesoftware.com/wiki/Category:List_of_Shader_Parameters>`__ and `proxies <https://developer.valvesoftware.com/wiki/Material_proxies>`__ to dictate how `shaders <https://developer.valvesoftware.com/wiki/Shader>`__ will show the game. We will go into greater detail later.

Editing FX that use VTFs
------------------------

A lot of fx in Titanfall use vtf's as textures. Therefore, if the corresponding vtf can be found, we can do almost anything with the fx's appearence.
Example Mod: `Exrill's Blue L-Star <https://northstar.thunderstore.io/package/EXRILL/Exrills_Blue_Lstar/>`_.
Since the L-Star has a physical bullet that is counted as fx, we can edit how it looks.

VTF Skins
---------

> Since vtf modding was majorly for introducing custom weapon skins. Most of the development on VTF modding was designed focused on custom weapon skins. Although understanding how custom weapon VTF skins is specific, it is needed in order to VTF mod other textures. 

Reviewer: ```suggestion
Since vtf modding was originally for introducing custom weapon skins, most of the development on it was focused on that. The same concepts apply to modding other textures though.
```

Most textures in the game use .dds but we can make them use .vtf. 

> What we'll be doing is extracting the model. Replacing the texture path in the mdl to lead to our texture. Then using our texture and allowing for extra properties.

Reviewer: ```suggestion
What we'll be doing is:

  - Extracting the model.
  - Replacing the texture path in the mdl to point to our texture.
  - Allowing for extra properties on our texture.
```
not sure if the last point is what you meant

Author: It isn't. Although I wrote it horribly, I meant something along the lines of, Creating a VMT file to manage our textures.

.. _VPK Tool: https://github.com/Wanty5883/Titanfall2/blob/master/tools/Titanfall_VPKTool3.4_Portable.zip

VPK Tool
--------
.. _cra0 VPKTool: https://github.com/Wanty5883/Titanfall2/blob/master/tools/Titanfall_VPKTool3.4_Portable.zip

.. _Harmony VPKTool: https://github.com/harmonytf/HarmonyVPKTool

You have 2 options for a VPK tool. Pick either the older VPK tool: `cra0 VPKTool`_ or the Newer VPK tool: `Harmony VPKTool`_ (better).

With your VPK tool opened. 'Open' ``englishclient_mp_common.pak000_dir.vpk`` which is located in ``Titanfall2/vpk``. 

> Inside of the VPK. Not all guns are named the same in files as they are in game. Here is `list of weapon names <https://noskill.gitbook.io/titanfall2/documentation/file-location/weapon/weapon-model>`_ to help you out. 

Reviewer: ```suggestion
Inside of the VPK, not all guns filenames match their ingame names. Here is `list of weapon names <https://noskill.gitbook.io/titanfall2/documentation/file-location/weapon/weapon-model>`_ to help you out. 
```

Navigate to ``models/weapons/car101``. Extract all the viewmodel versions (ptpov) and normal model (w) mdl's.

Hex Editor
----------

To change the path in the .mdl to the custom .vmt. We need a hex editor. I will use `HxD`_, but you can also use `ida`_, or anything else; its personal preference. 

.. _HxD: https://mh-nexus.de/en/hxd/
.. _ida: https://hex-rays.com/ida-free/

> In your hex editor. Open your .mdl. We need to get closer to the string we need or else you'll be scrolling and searching for hours. Search:(CTRL+F) for skin_31. If you don't get any matches, try skn_31, skin31, elite, or prime.

Reviewer: ```suggestion
Open your .mdl in your hex editor. We need to get closer to the string we need or else you'll be scrolling and searching for hours. Search:(CTRL+F) for skin_31. If you don't get any matches, try skn_31, skin31, elite, or prime.
```

> The path should look something like ``.models\Weapons_R2\weaponname\weaponname_skin_31``. Before this skin_31 path. We should find an almost exact path but without ``skin_31``. This is the path we want. This non-skin_31 path is the path to the default weapon skin.

Reviewer: Theres a bunch of grammar mistakes here, and im not quite sure what its trying to say. which path is meant? the one shown or one without the skin31 part?

Author: I thought it was clear and adds addition insight. I'll ask other to see if it's also unclear to them. Then consider changes.

Reviewer 2: Open the .mdl in a hex editor. We need to get closer to the string we need, otherwise you'll be scrolling and searching for hours. Search (CTRL+F) for skin_31. If there are no matches, try: skn_31, skin31, elite or prime.

> Now, before you edit, you have to realize hex editors are just built different (cant help it). You cannot add or delete data with a hex editor, only replace it. 

Reviewer: ```suggestion
Before editing you need to be aware that hex editors cannot add or delete data, only replace it. 
```

> I recommend changing only the name towards the end of the path. We'll go from ``.models\Weapons_r2\car_smg\CAR_smg`` to ``.models\weapons_r2\car_smg\car_ctm``. Note the capitalization. Some vpk repacking tools cannot repack properly if we edit in captials. Now do these changes for ``ptpov_`` and/or ``w_`` model(s). 

Reviewer: ```suggestion
I recommend changing only the last section of the path. We'll change``.models\Weapons_r2\car_smg\CAR_smg`` to ``.models\weapons_r2\car_smg\car_ctm``. Note the capitalization, some vpk repacking tools cannot repack properly if the changed path contains captials. Now do these changes for ``ptpov_`` and/or ``w_`` model(s). 
```

Reviewer: maybe a link to the ptpov_ and w_ path explanation

Creating VMT
-------------

In the same folder you extracted your mdl's. Make a ``materials`` folder next to the ``models`` folder. 

> Inside the ``materials`` folder. Recreate the path you hex edited, but the last name is a .vmt file. Not a folder. Our path of folders is ``.models/weapons_r2/car_smg``, our .vmt file would be named ``car_ctm.vmt``. 

Reviewer: ```suggestion
Recreate the path you change in the ``materials`` folder, such that the last section is a .vmt file. 

::

	materials
	└─ models
	      └─ weapons_r2
	            └─ car_smg
	                 └─ car_ctm.vmt
```

		"$nocull" "1"
	}

> When we use vtf textures, we can only use the albedo and normal. Learn more about `texture maps <https://retryy.gitbook.io/tf2/wiki/create/texturemaps>`_ here.

Reviewer: ```suggestion
When we use vtf textures, we can only use the albedo and normal texture maps, learn more about `here <https://retryy.gitbook.io/tf2/wiki/create/texturemaps>`_.
```

VTFEdit
--------

`VTFEdit`_ is a tool to edit, view, and create .vtf files.

.. _VTFEdit: https://nemstools.github.io/pages/VTFLib-Download.html

Launch `VTFEdit`_. Top left, click ``File``, ``Import``, find and Import your custom texture(s). 

When importing your normal map. Choose to import as a ``Volume Map``
When importing your diffuse map. Choose to import as a ``Animated Map``

More info about .vtf format possibilities `here <https://retryy.gitbook.io/tf2/wiki/create/formats>`_, or the official source docs `here <https://developer.valvesoftware.com/wiki/Valve_Texture_Format>`_.

After that, save your new .vtf's into the same folder as your custom .vmt with a simple name.

Configuring your .vmt
---------------------

> To know which texture/texture map you have. Analyze `this <https://retryy.gitbook.io/tf2/wiki/create/texturemaps>`_ wiki.

Reviewer: ```suggestion
It's highly recommende to read `this <https://retryy.gitbook.io/tf2/wiki/create/texturemaps>`_ wiki to understand what texture maps you might want.
```

> In the ``"$basetexture"`` argument. You'll enter your .vtf texture directory. We'll use ``models\weapons_r2\car_ctm\NAMEOFVTF``. This is pointing to your custom diffuse .vtf with the simple name. Do not add the ``.vtf`` file extension to the end of the directory; ``models\weapons_r2\car_ctm\NAMEOFVTF.vtf``, or the game cannot find the texture.
> 
> Do the same for your normal map in the ``"$bumpmap"`` argument.
> 
> Now create another vtf with literally any image. Put its path in ``"$texture2"`` argument. As far as i know, this is necessary even though the texture isn't used. (i never had to do this but maybe you need it)

Reviewer: ```suggestion
In the ``"$basetexture"`` argument enter your .vtf texture directory. We'll use ``models\weapons_r2\car_ctm\NAMEOFVTF``. This should point to your custom diffuse .vtf with the simple name. The game expects these paths to be without the ``.vtf`` file extension - don't add it.

Do the same for adding your normal map with the ``"$bumpmap"`` argument.

In some cases you might have to create another vtf with literally any image. Put its path in the ``"$texture2"`` argument. As far as i know, this is sometimes necessary even though the texture isn't used.
```

A example of the finaly .vtm here might be needed, or a re-wording of the "arguments". Maybe using more common JSON terminology of keys?

Final VPK folder
----------------

Your root folder should look somewhat like this

::

	root
	├─ materials
	│  └─ models
	│     └─ weapons_r2
	│        └─ car_smg
	│           ├─ YOURTEXTURE.vtf
	│           ├─ YOURTEXTURE.vtf
	│           └─ car_ctm.vmt
	└─ models
	   └─ weapons
	      └─car101
	        ├─ ptpov_car101.mdl
	        └─ w_car101.mdl

Finished.
---------

You're done! You just need to pack it into a vpk with the vpk tool (for our gun mod, we'd repack to ``englishclient_mp_common.pak000_dir.vpk``), and put it into a northstar mod. 

> Help with repacking `here <https://noskill.gitbook.io/titanfall2/intro/duction/vpk-packpack>`_, and help with Northstar mods `here <https://r2northstar.readthedocs.io/en/latest/guides/gettingstarted.html>`_.

Reviewer: This should specify where in the northstar mod folder structure the vpk should go

Author: What? Like putting it into a Northstar mod? or packing it into a vpk?

Making your Skin Animated
-------------------------

> To add animation functionality, all we need to do is add a proxie and change our albedo vtf. 

Reviewer: explain what a proxie is and how/where we add one.

Author: I don't know what they are. So I will learn and then add that.
